### ISSUE TITLE
Ability to provide additional fields to append to json format

### ISSUE THREAD

**Issue by mradzevicius** [spring-cloud/spring-cloud-gcp#1310]
# Description

Ability to log additional context variables provides ability to search logs sent to Stackdriver without using MDC context to do that. Something similar to `StructuredArguments` in `logstash-logback-encoder`.

ie:
```java
logger.info("job {} has failed", kv("job_id", job.getId(), kv("retry_count", job.getRetryCount()))
```

after logging fields would be appended to the json object which is sent to stdout: 

```json
{
  "job_id": "xxxx-yyy-zzzz",
  "retry_count": 3
}
```

Then those fields would be logged into stackdriver when using Logging view then we could filter out by additional parameters either `job_id` or `retry_count`. 

Maybe it is already possible and I just did not find a way how to do this?

**Comment by meltsufin** [comment]
Have you looked at the `includeMDC` [property](https://docs.spring.io/spring-cloud-gcp/docs/1.1.0.BUILD-SNAPSHOT/reference/htmlsingle/#_log_via_console)? Does it address your requirements?

**Comment by mradzevicius** [comment]
@meltsufin yes i've looked at it, unfortunately not in this case - it has its place for usage.

Also a bonus, would be ability to configure `serviceContext` property or an ability to add static attributes (that might be possible via out of the box logback):

```json
{
"serviceContext": {"service": "app-name", "version": "app-version" }
}
```

That would also benefit Stackdriver error reporting when exceptions are thrown and catched also filtering by app/version when you have multiple running. By specifying `serviceContext`, additional information appears in Error Reporting window in GCP with those values which are also useful, otherwise exception appears in GKE case just as a general "seen in gke-instances" versus "seen in app-name:app-version". 

I've managed to achieve the above logging and the serviceContext concerns with using `logstash-logback-encoder` using stackdriver logging output as a reference to craft it manually. ie:

logback config:
```
...
<appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
            <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
                <providers>
                    <mdc/>
                    <provider class="logging.Sleuth"></provider>
                    <logstashMarkers/>
                    <arguments/>
                    <pattern>
                        <pattern>
                            {
                            "serviceContext": {"service": "${appArtifactId}", "version": "${appVersion}" },
                            "severity": "%level",
                            "thread": "%thread",
                            "logger": "%logger{36}.%M",
                            "message" : "%message %stack{30,30,2048,rootFirst}"
                            }
                        </pattern>
                    </pattern>
                </providers>
            </encoder>
        </appender>
...
```

The `Sleuth` provider, does almost same thing as `org.springframework.cloud.gcp.autoconfigure.logging.StackdriverJsonLayout` in terms for tracing requirements in json format to be correctly digested by Stackdriver Trace.

Don't get me wrong - now stackdriver logging support within spring boot is great , from my point of view -  its these couple enhancements if they would come out of the box would provide better experience and a bit closer integration with Stackdriver product stack. 

Also the above findings are from personal trial-error and poking around stackdriver & inspecting `google-fluentd` to understand and to make it work - it might not be the best approach, but at the moment it works. I just wish it would be a bit easier to configure this behavior via spring integration instead of finding other ways. 

[1] Regarding `serviceContext` https://cloud.google.com/error-reporting/docs/formatting-error-messages

**Comment by katholen** [comment]
We are trying to enable the same thing on our end. I agree with @mradzevicius that it would be ideal if this was configurable via the core implementation. We will try out your solution and let you know if it works.

**Comment by mradzevicius** [comment]
@meltsufin sorry for radio silence. I think it could work with `includeMDC`, however as this field is part of supported fields and how stackdriver functinos moreover its static I think it would be easier to have it supported via library (ie. property configuration), instead of the user having to set the `serviceContext` field manually via mdc.

**Comment by meltsufin** [comment]
@mradzevicius Thanks for the update. Can you explain more on how we could make this work with Logback, or would we need a different logging library?
